File: apps/macos/Sources/BridgeVMProductE2E/T17TerminalReportTail.swift
import Foundation

/// Recognizes the final host report and only the bounded audio shutdown tail.
enum T17TerminalReportTail {
    static let maxTailBytes = 4 * 1024
    static let maxTailLines = 16
    private static let footer = Data("\n--- end ---\n".utf8)
    private static let hostLines = [
        #"^hda CoreAudio callback enqueue: state=stopping reason=[a-z-]+ osstatus=-?[0-9]+ expected=(true|false)$"#,
        #"^hda CoreAudio lifecycle: operation=(stop|dispose) osstatus=-?[0-9]+ success=(true|false)$"#,
        #"^hda CoreAudio stats: [a-z][a-z0-9_]*=[0-9]+( [a-z][a-z0-9_]*=[0-9]+)*$"#,
    ]

    static func isComplete(rawSuffix: Data, nonce: String, generation: UInt64) -> Bool {
        let ack = Data("HOST-DIAGNOSTIC-STOP: generation=\(generation) nonce=\(nonce) request consumed; ending run through final report".utf8)
        let serialMarker = Data("\n--- serial (tail) ---\n".utf8)
        let bannerMarker = Data("\n=== EDK2 boot probe (with Apple hv_gic) ===\n".utf8)
        let stopMarker = Data("stop: host diagnostic stop requested\n".utf8)
        guard let ackRange = rawSuffix.range(of: ack),
              (ackRange.lowerBound == 0 || rawSuffix[ackRange.lowerBound - 1] == 10),
              ackRange.upperBound < rawSuffix.count, rawSuffix[ackRange.upperBound] == 10,
              let serial = rawSuffix.range(of: serialMarker, in: ackRange.upperBound..<rawSuffix.count),
              let banner = rawSuffix.range(of: bannerMarker, in: ackRange.upperBound..<serial.lowerBound),
              let stop = rawSuffix.range(of: stopMarker, in: banner.upperBound..<serial.lowerBound) else {
            return false
        }
        let countStart = rawSuffix.range(of: Data([10]), options: .backwards,
                                         in: stop.upperBound..<serial.lowerBound)?.upperBound ?? stop.upperBound
        guard let countLine = String(data: rawSuffix[countStart..<serial.lowerBound], encoding: .ascii),
              let count = serialCount(countLine),
              count.output <= rawSuffix.count - serial.upperBound else { return false }
        let footerStart = serial.upperBound + count.output
        guard rawSuffix.count - footerStart >= footer.count,
              rawSuffix[footerStart..<footerStart + footer.count] == footer,
              let displayed = String(data: rawSuffix[serial.upperBound..<footerStart], encoding: .utf8),
              (!count.legacy || !displayed.contains("\u{FFFD}")) else { return false }
        let tail = rawSuffix[footerStart + footer.count..<rawSuffix.count]
        guard let text = String(data: tail, encoding: .utf8) else { return false }
        return isBoundedHostShutdownTail(text)
    }

    private static func serialCount(_ line: String) -> (output: Int, legacy: Bool)? {
        let modern = "serial raw bytes: "
        if line.hasPrefix(modern) {
            let fields = String(line.dropFirst(modern.count)).components(separatedBy: " output bytes: ")
            guard fields.count == 2, let raw = decimal(fields[0]), let output = decimal(fields[1]),
                  output >= raw else { return nil }
            return (output, false)
        }
        let old = "serial bytes: "
        guard line.hasPrefix(old), let output = decimal(String(line.dropFirst(old.count))) else { return nil }
        return (output, true)
    }

    private static func decimal(_ value: String) -> Int? {
        let bytes = value.utf8
        guard !bytes.isEmpty, bytes.count <= 8,
              bytes.allSatisfy({ (48...57).contains($0) }) else { return nil }
        return Int(value)
    }

    private static func isBoundedHostShutdownTail(_ tail: String) -> Bool {
        if tail.isEmpty { return true }
        guard tail.utf8.count <= maxTailBytes, tail.hasSuffix("\n") else { return false }
        let lines = tail.split(separator: "\n", omittingEmptySubsequences: false)
        guard lines.count <= maxTailLines + 1, lines.last?.isEmpty == true else { return false }
        return lines.dropLast().allSatisfy { line in
            let value = String(line)
            return hostLines.contains { value.range(of: $0, options: .regularExpression) != nil }
        }
    }
}

File: scripts/live-gates/t17_terminal_report_tail.py
"""Recognize an authenticated T17 terminal report with bounded host audio tail."""
from __future__ import annotations

import re

ACK = re.compile(rb"^HOST-DIAGNOSTIC-STOP: generation=([0-9]+) nonce=([0-9a-f]{32}) request consumed; ending run through final report$", re.M)
COMPLETE_DETAIL = re.compile(r"host_stop=status=complete,generation=([0-9]+),nonce=([0-9a-f]{32}),report=complete,helper=terminal,log_offset=([0-9]+)(?:;|\Z)")
MAX_AUDIO_TAIL_BYTES = 4 * 1024
MAX_AUDIO_TAIL_LINES = 16
FOOTER = b"\n--- end ---\n"
SERIAL = b"\n--- serial (tail) ---\n"
BANNER = b"\n=== EDK2 boot probe (with Apple hv_gic) ===\n"
STOP = b"\nstop: host diagnostic stop requested\n"
COUNT_NEW = re.compile(rb"(?:^|\n)serial raw bytes: ([0-9]{1,8}) output bytes: ([0-9]{1,8})\Z")
COUNT_LEGACY = re.compile(rb"(?:^|\n)serial bytes: ([0-9]{1,8})\Z")
AUDIO_LINES = (
    re.compile(r"hda CoreAudio callback enqueue: state=stopping reason=[a-z-]+ osstatus=-?[0-9]+ expected=(true|false)"),
    re.compile(r"hda CoreAudio lifecycle: operation=(stop|dispose) osstatus=-?[0-9]+ success=(true|false)"),
    re.compile(r"hda CoreAudio stats: [a-z][a-z0-9_]*=[0-9]+( [a-z][a-z0-9_]*=[0-9]+)*"),
)


def bounded_host_shutdown_tail(tail: str) -> bool:
    if not tail:
        return True
    if len(tail.encode("utf-8")) > MAX_AUDIO_TAIL_BYTES or not tail.endswith("\n"):
        return False
    lines = tail[:-1].split("\n")
    return 1 <= len(lines) <= MAX_AUDIO_TAIL_LINES and all(
        any(pattern.fullmatch(line) for pattern in AUDIO_LINES) for line in lines
    )


def terminal_report(tail: bytes, detail: str, tail_offset: int) -> tuple[int, str] | None:
    claim = COMPLETE_DETAIL.search(detail)
    if claim is None or int(claim[3]) < tail_offset or int(claim[3]) > tail_offset + len(tail):
        return None
    raw = tail[int(claim[3]) - tail_offset:]
    ack = next((match for match in ACK.finditer(raw) if match[2] == claim[2].encode()), None)
    if ack is None or ack[1] != claim[1].encode():
        return None
    serial = raw.find(SERIAL, ack.end())
    banner = raw.find(BANNER, ack.end(), serial)
    stop = raw.find(STOP, banner, serial)
    if serial < 0 or banner < 0 or stop < 0:
        return None
    header = raw[stop + len(STOP):serial]
    modern = COUNT_NEW.search(header)
    legacy = modern is None
    if modern is not None:
        raw_size, size = int(modern[1]), int(modern[2])
        if size < raw_size:
            return None
    else:
        old = COUNT_LEGACY.search(header)
        if old is None:
            return None
        size = int(old[1])
    footer = serial + len(SERIAL) + size
    if footer > len(raw) or raw[footer:footer + len(FOOTER)] != FOOTER:
        return None
    try:
        displayed = raw[serial + len(SERIAL):footer].decode("utf-8")
        suffix = raw[footer + len(FOOTER):].decode("ascii")
    except UnicodeDecodeError:
        return None
    if (legacy and "\ufffd" in displayed) or not bounded_host_shutdown_tail(suffix):
        return None
    return int(claim[1]), raw[stop:serial].decode("utf-8", errors="replace")
